### ISSUE TITLE
Data corruption

### ISSUE THREAD

**Issue by acehoss** [acehoss/rnsh#20]
The latest changes in RNS have fixed the issues with channels and listeners unresponsive. I ran `top` with `rnsh` overnight over both ethernet and LoRa and the sessions are still running this morning--something that was not possible on the previous version.

However, I am noticing some subtle problems. In the session running over ethernet, I'm noticing text out of place after a minute or so. I simultaneously ran a session over SSH and it had no problems. (Screenshots at the bottom)

It seems like macOS `top` doesn't clear the terminal or the line, but rather only updates characters when they change by positioning the cursor and printing the updated text. This means if a data message is lost or duplicated, it could result in this garbling (particularly if relative cursor positioning is used). This could be a bug introduced by the recent changes in RNS, but I think I may have seen issues like this before that update.

SSH:
![macos-top-ssh](https://github.com/acehoss/rnsh/assets/5148966/fe1540ea-24dd-4440-a70d-ae0cd3417e62)

rnsh - notice the garbled first few lines compared to the SSH session:
![macos-top-rnsh](https://github.com/acehoss/rnsh/assets/5148966/38f3c86d-fb8a-40bb-98cc-0665ed71423d)

**Comment by acehoss** [comment]
An even more garbed `top` over `rnsh` session after about 20 minutes. The `top` over `ssh` session is still running and has no corruption.

![CleanShot 2023-05-13 at 04 34 39@2x](https://github.com/acehoss/rnsh/assets/5148966/02c4cf3b-6e2f-4ac6-a2cb-f0704aa69e65)

**Comment by acehoss** [comment]
I ran a test that printed a sequence of numbers, copied the terminal output into a file and compared the text to a locally generated file. There appear to be some out-of-order segments in the received text.

*Text Generator* `i=0; while [ $i -lt 1000000 ]; do i=$((i+1)); printf "%032d\n" $i; done`

```
$ diff -y --suppress-common-lines over-rnsh.txt locally-generated.txt                      
00000000000037510						|       00000000000000000000000000037498
								>       00000000000000000000000000037499
								>       00000000000000000000000000037500
								>       00000000000000000000000000037501
								>       00000000000000000000000000037502
								>       00000000000000000000000000037503
								>       00000000000000000000000000037504
								>       00000000000000000000000000037505
								>       00000000000000000000000000037506
								>       00000000000000000000000000037507
								>       00000000000000000000000000037508
								>       00000000000000000000000000037509
								>       00000000000000000000000000037510
0000000000000000000000037498					|       00000000000000000000000000037523
00000000000000000000000000037499				<       
00000000000000000000000000037500				<       
00000000000000000000000000037501				<       
00000000000000000000000000037502				<       
00000000000000000000000000037503				<       
00000000000000000000000000037504				<       
00000000000000000000000000037505				<       
00000000000000000000000000037506				<       
00000000000000000000000000037507				<       
00000000000000000000000000037508				<       
00000000000000000000000000037509				<       
000000000000000000000000000000000000000000000037523		<       
000000000000000000000000000000000000000000000037834		|       00000000000000000000000000037821
00000000000000000000000000037835				<       
00000000000000000000000000037836				<       
00000000000000000000000000037837				<       
00000000000000000000000000037838				<       
00000000000000000000000000037839				<       
00000000000000000000000000037840				<       
00000000000000000000000000037841				<       
00000000000000000000000000037842				<       
00000000000000000000000000037843				<       
00000000000000000000000000037844				<       
00000000000000000000000000037845				<       
0000000000000000000000037821					<       
00000000000037846						|       00000000000000000000000000037834
								>       00000000000000000000000000037835
								>       00000000000000000000000000037836
								>       00000000000000000000000000037837
								>       00000000000000000000000000037838
								>       00000000000000000000000000037839
								>       00000000000000000000000000037840
								>       00000000000000000000000000037841
								>       00000000000000000000000000037842
								>       00000000000000000000000000037843
								>       00000000000000000000000000037844
								>       00000000000000000000000000037845
								>       00000000000000000000000000037846
```

**Comment by markqvist** [comment]
Just FYI, I am trying to debug/diagnose this now.

**Comment by markqvist** [comment]
Ok, I'm a bafoon. When I added the windowing stuff, I "solved" a lock hang by adding a thread decoupler at the wrong place in the call chain, causing a potential race condition for the callback handling the extension of the buffer data.

Fix incoming :)

**Comment by acehoss** [comment]
I prefer by a wide margin a one-line oopsie over an unforeseen architecture issue requiring an extensive rewrite. 😅

**Comment by markqvist** [comment]
Haha, exactly! Well, I just pushed some commits that should fix the issue. Let me know when you get some time to test it out!

I've tested it out both with:
- A small script reading a file from disk and sending it over a buffer. Comparing afterwards with diff is an exact match this time around :)
- Running `rnsh` sessions and checking for any garbling over time. None so far. Even had four instances of `top` running inside `tmux` for a couple of hours, and all looked good.
- Using `rnsh` to `cat` various files and comparing the output with `diff`. All checks out :)
- Tested both over local instance, LAN and multi-hop over the internet.

Fingers crossed it's good this time around :)
